frame/support/procedural/src/compact_pallet_error.rs

```diff
@@ -0,0 +1,116 @@
+	let syn::DeriveInput { ident: name, generics, data, .. } = match syn::parse(input) {
+		Ok(input) => input,
+		Err(e) => return e.to_compile_error().into(),
+	};
+
+	let frame_support = match generate_crate_access_2018("frame-support") {
+		Ok(c) => c,
+		Err(e) => return e.into_compile_error().into(),
+	};
+	let frame_support = &frame_support;
+	let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+	let compactness_check = match data {
+		syn::Data::Struct(syn::DataStruct { struct_token, fields, .. }) => {
+			if fields.len() > 1 {
+				let msg = "Cannot derive `CompactPalletError` for structs with more than 1 field";
+				return syn::Error::new(struct_token.span, msg).into_compile_error().into()
+			}
+
+			match fields {
+				syn::Fields::Named(mut f) if f.named.len() == 1 => {
+					let field_ty = f.named.pop().unwrap().into_value().ty;
+					quote::quote! {
+						<
+							#field_ty as #frame_support::traits::CompactPalletError
+						>::check_compactness()
+					}
+				},
+				syn::Fields::Unnamed(mut f) if f.unnamed.len() == 1 => {
+					let field_ty = f.unnamed.pop().unwrap().into_value().ty;
+					quote::quote! {
+						<
+							#field_ty as #frame_support::traits::CompactPalletError
+						>::check_compactness()
+					}
+				},
+				_ => quote::quote!(true),
+			}
+		},
+		syn::Data::Enum(syn::DataEnum { variants, .. }) => {
+			let field_tys = variants
+				.into_iter()
+				.map(|variant| match variant.fields {
+					syn::Fields::Named(mut f) if f.named.len() == 1 =>
+						Ok(Some(f.named.pop().unwrap().into_value().ty)),
+					syn::Fields::Unnamed(mut f) if f.unnamed.len() == 1 =>
+						Ok(Some(f.unnamed.pop().unwrap().into_value().ty)),
+					syn::Fields::Unit => Ok(None),
+					_ if variant.ident == "__Ignore" => Ok(None),
+					_ => {
+						let msg = "Cannot derive `CompactPalletError` for enum with variants \
+						containing more than 1 field";
+						let err = syn::Error::new(variant.ident.span(), msg);
+						Err(err)
+					},
+				})
+				.collect::<Result<Vec<Option<syn::Type>>, syn::Error>>();
+
+			let field_tys = match field_tys {
+				Ok(tys) => tys.into_iter().filter_map(identity).collect::<Vec<_>>(),
+				Err(e) => return e.to_compile_error().into(),
+			};
+
+			if field_tys.is_empty() {
+				quote::quote!(true)
+			} else {
+				quote::quote! {
+					#(
+						<
+							#field_tys as #frame_support::traits::CompactPalletError
+						>::check_compactness()
+					)&&*
+				}
+			}
+		},
+		syn::Data::Union(syn::DataUnion { union_token, .. }) => {
+			let msg = "Cannot derive `CompactPalletError` for union; please implement it directly";
+			return syn::Error::new(union_token.span, msg).into_compile_error().into()
+		},
+	};
+
+	quote::quote!(
+		const _: () = {
+			impl #impl_generics #frame_support::traits::CompactPalletError
+				for #name #ty_generics #where_clause
+			{
+				fn check_compactness() -> bool {
+					#compactness_check
```

Comment:
I'm still not sure why we need a runtime check, why implementing CompactPalletError isn't enough to ensure that the type is a CompactPalletError ?

Instead of having a runtime check which will be trigger only when the user implement the test in construct_runtime maybe it is better to do static check by saying a type is CompactPalletError when it implement the trait.

Then this derive macro would just add some bound:
```
impl #impl_generics #frame_support::traits::CompactPalletError
				for #name #ty_generics #where_clause #( , #tys: CompactPalletError )* {}
				// actually it is better to add into the where clause variable directly but this snippet is an example.
```

Reply:
There already is a static check in fact, my reasoning to also have a runtime check is to ensure that people don't actually blindly derive `CompactPalletError` and expect it work -- they'll actually have to implement `check_compactness` and prove that the encoded size is 1 byte.

Reply:
yes indeed, then I'm ok for having both.